x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/ratelimiter.go

```diff
@@ -16,33 +16,66 @@ import (
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
-type RateLimiter map[string]*rate.Limiter
+type RateLimiter struct {
+	window     time.Duration
+	fixedLimit *int
+	byEndpoint map[string]endpointRateLimiter
+}
+
+type endpointRateLimiter struct {
+	limiter *rate.Limiter
+	ready   chan struct{}
+}
 
-func NewRateLimiter() RateLimiter {
-	r := make(RateLimiter)
-	return r
+const waitDeadline = 30 * time.Minute
+
+func NewRateLimiter(window time.Duration, fixedLimit *int) *RateLimiter {
+	endpoints := make(map[string]endpointRateLimiter)
+	r := RateLimiter{
+		window:     window,
+		fixedLimit: fixedLimit,
+		byEndpoint: endpoints,
+	}
+	r.fixedLimit = fixedLimit
+	return &r
 }
 
-func (r RateLimiter) limiter(path string) *rate.Limiter {
-	if existing, ok := r[path]; ok {
+func (r RateLimiter) endpoint(path string) endpointRateLimiter {
+	if existing, ok := r.byEndpoint[path]; ok {
 		return existing
 	}
-	initial := rate.NewLimiter(1, 1) // Allow a single fetch operation to obtain limits from the API
-	r[path] = initial
-	return initial
+	limit := rate.Limit(1)
+	if r.fixedLimit != nil {
+		limit = rate.Limit(float64(*r.fixedLimit) / r.window.Seconds())
+	}
+	limiter := rate.NewLimiter(limit, 1) // Allow a single fetch operation to obtain limits from the API
+	ready := make(chan struct{})
+	close(ready)
```

Comment:
Suggest something like
```
// alwaysReady is an non-blocking chan.
var alwaysReady = make(chan struct{})

func init() { close(alwaysReady) }

func (r RateLimiter) endpoint(path string) endpointRateLimiter {
	if existing, ok := r.byEndpoint[path]; ok {
		return existing
	}
	limit := rate.Limit(1)
	if r.fixedLimit != nil {
		limit = rate.Limit(float64(*r.fixedLimit) / r.window.Seconds())
	}
	limiter := rate.NewLimiter(limit, 1) // Allow a single fetch operation to obtain limits from the API
	newEndpointRateLimiter := endpointRateLimiter{
		limiter: limiter,
		ready:   alwaysReady,
	}
	r.byEndpoint[path] = newEndpointRateLimiter
	return newEndpointRateLimiter
}
```
so that we have a single never-blocking chan.

Reply:
Done with the name `immediatelyReady`.
It's true that channel will always be ready, but the `endpointRateLimiter`'s `ready` field won't always have that value...
I don't know. Is this better, or the same? (or less good?)